Trail-of-bits #13

pkhivesara opened the pull request:
Fixes the following issue: https://github.com/OpenArchive/OA-Trail-of-Bits/issues/13

tladesignz commented:
Both `space.host` and `space.name` look pretty weird in this context. Does `space.name` contain a full URL, including scheme and path? 

Also, string concatenation `val url = "$spaceName/$username"` looks dangerous. Shouldn't there be some URL mangling class be involved, to make sure the URL is invalid, afterwards? If $username is completely controlled by the user, this could also lead to some annoying bugs/security issues.

Maybe this should refactored to something else then `spaceName` resp. `space.name` if it contains a full URL? If not, then how do the other URL components come in? Via the `OkHttpBaseClient` or the `Request.Builder`? Does that clean up a potentially botched `Request.url` argument?


In general, I'm quite confused about what Save Android does here.

In the iOS version for a WebDAV-based server, the user would just enter the domain of a e.g. Nextcloud server like `nextcloud.example.com` and a [URL formatter](https://github.com/OpenArchive/Save-app-ios/blob/master/Shared/Utils/Formatters.swift#L135-L179) would immediately clean this up to `https://nextcloud.example.com/remote.php/webdav/`. (which could then be further edited by the user to adapt to a completely different WebDAV server.)
Credentials will be added as HTTP "basic authentication" headers. If a connection attempt succeeds, the URL is stored and that's it.

Can you investigate, what the Android version does here? Why does it place the username in the URL? Suggesting by the (now deleted) hardcoded URL mangling, and some added headers, it seems to use a completely different interface? Is that a compatible WebDAV approach or does this just work with OwnCloud/Nextcloud?

pkhivesara commented:
> Both `space.host` and `space.name` look pretty weird in this context. Does `space.name` contain a full URL, including scheme and path?
> 
> Also, string concatenation `val url = "$spaceName/$username"` looks dangerous. Shouldn't there be some URL mangling class be involved, to make sure the URL is invalid, afterwards? If $username is completely controlled by the user, this could also lead to some annoying bugs/security issues.
> 
> Maybe this should refactored to something else then `spaceName` resp. `space.name` if it contains a full URL? If not, then how do the other URL components come in? Via the `OkHttpBaseClient` or the `Request.Builder`? Does that clean up a potentially botched `Request.url` argument?
> 
> In general, I'm quite confused about what Save Android does here.
> 
> In the iOS version for a WebDAV-based server, the user would just enter the domain of a e.g. Nextcloud server like `nextcloud.example.com` and a [URL formatter](https://github.com/OpenArchive/Save-app-ios/blob/master/Shared/Utils/Formatters.swift#L135-L179) would immediately clean this up to `https://nextcloud.example.com/remote.php/webdav/`. (which could then be further edited by the user to adapt to a completely different WebDAV server.) Credentials will be added as HTTP "basic authentication" headers. If a connection attempt succeeds, the URL is stored and that's it.
> 
> Can you investigate, what the Android version does here? Why does it place the username in the URL? Suggesting by the (now deleted) hardcoded URL mangling, and some added headers, it seems to use a completely different interface? Is that a compatible WebDAV approach or does this just work with OwnCloud/Nextcloud?

Completely understand the confusion here. I was chasing a rabbit hole as well yesterday on this. The older (now deleted) approach only worked for WebDAV providers like NextCloud/Hetnzer.

If a user decided to add a arbitrary webdav provider that they are self hosting, the old approach would fail. 

I switched it over to use the ``spaceName`` and ``username`` to make it work in tangent with what is shown within the web app for nextcloud/hetzner (There is a section which points to what URL to use to login via WebDav protocol from mobile apps). ``spaceName`` has the base URL only.

In Android, the ``space.host`` ended up being the scheme, host and the path (``https://nextcloud.example.com/remote.php/webdav/``), which would obviously fail because that is not the correct URL to try logging in a user. 

The expected URL for login is supposed to be: ``https://nextcloud.example.com/yourusername``

I am not entirely sure why and how is iOS able to login a user with ``https://nextcloud.example.com/remote.php/webdav/``? When I try the same URL on Android, I get a 404 error code.


>  If not, then how do the other URL components come in? Via the OkHttpBaseClient or the Request.Builder? 

Headers are added via ``RequestBuilder`` and the username, password is appended using ``OkHttpBaseClient``. At the end if the response if a success, only then do we store the response body.

tladesignz commented:
@pkhivesara, thanks for the explanation!

> I switched it over to use the `spaceName` and `username` to make it work in tangent with what is shown within the web app for nextcloud/hetzner (There is a section which points to what URL to use to login via WebDav protocol from mobile apps). `spaceName` has the base URL only.

This should seriously be refactored to something like `baseUrl`.

> In Android, the `space.host` ended up being the scheme, host and the path (`https://nextcloud.example.com/remote.php/webdav/`), which would obviously fail because that is not the correct URL to try logging in a user.

This should also be refactored to something like `fullUrl`. Then `baseUrl` should not be stored independently, but should be a method which actually strips `fullUrl` to the base using a some URL/URI class which can handle this correctly.

> The expected URL for login is supposed to be: `https://nextcloud.example.com/yourusername`

Login? I don't need a "login" with my Nextcloud server. I just send the `Authorization` header along with every request, which contains a HTTP Basic Auth BASE64-encoded username/password combo.

Does the Android version need to go to that username address for a "login" to acquire a session token?

> I am not entirely sure why and how is iOS able to login a user with `https://nextcloud.example.com/remote.php/webdav/`? When I try the same URL on Android, I get a 404 error code.

If you do a GET there with proper authorization, it should return the listing for the root folder.
If I don't add an authorization header, I'll get this:

```
* Preparing request to https://nextcloud.benjaminerhart.com/remote.php/webdav/
* Current time is 2023-01-04T14:49:54.434Z
* Enable automatic URL encoding
* Saved 7 cookies
* Using default HTTP version
* Enable SSL validation
*   Trying 2a03:4000:b:217:c8aa:95ff:fe94:3544:443...
* Connected to nextcloud.benjaminerhart.com (2a03:4000:b:217:c8aa:95ff:fe94:3544) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN, server accepted to use http/1.1
* Server certificate:
*  subject: CN=nextcloud.benjaminerhart.com
*  start date: Nov 24 09:36:05 2022 GMT
*  expire date: Feb 22 09:36:04 2023 GMT
*  subjectAltName: host "nextcloud.benjaminerhart.com" matched cert's "nextcloud.benjaminerhart.com"
*  issuer: C=US; O=Let's Encrypt; CN=R3
*  SSL certificate verify ok.

> GET /remote.php/webdav/ HTTP/1.1
> Host: nextcloud.benjaminerhart.com
> User-Agent: insomnia/2022.6.0
> Accept: */*

* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* old SSL session ID is stale, removing
* Mark bundle as not supporting multiuse

< HTTP/1.1 401 Unauthorized
< Date: Wed, 04 Jan 2023 14:49:54 GMT
< Server: Apache/2.4.54 (Debian)
< Referrer-Policy: no-referrer
< X-Content-Type-Options: nosniff
< X-Frame-Options: SAMEORIGIN
< X-Permitted-Cross-Domain-Policies: none
< X-Robots-Tag: none
< X-XSS-Protection: 1; mode=block
< Set-Cookie: oclg6vjwjihc=08uhgm17h36u1qt4uhp8pi4g6j; path=/; secure; HttpOnly; SameSite=Lax
< Expires: Thu, 19 Nov 1981 08:52:00 GMT
< Cache-Control: no-store, no-cache, must-revalidate
< Pragma: no-cache
< Set-Cookie: oc_sessionPassphrase=E9M72YpHntUOKfIuojqpWhH3%2FSq8LhQMXYS98nTZlvdPmyC203s1g8c%2BiMUvzf7%2B6C2JsJT4gNk9JmjHIBwwFn7miuEWrm%2BBanOF82zd%2FvUmUcNuSZ6uFVl%2Fm2HkFpb%2F; path=/; secure; HttpOnly; SameSite=Lax

* Replaced cookie oclg6vjwjihc="15ij7urrbcqk2m9sgc393bgo5s" for domain nextcloud.benjaminerhart.com, path /, expire 0

< Set-Cookie: oclg6vjwjihc=15ij7urrbcqk2m9sgc393bgo5s; path=/; secure; HttpOnly; SameSite=Lax

* Replaced cookie oclg6vjwjihc="15ij7urrbcqk2m9sgc393bgo5s" for domain nextcloud.benjaminerhart.com, path /, expire 0

< Set-Cookie: oclg6vjwjihc=15ij7urrbcqk2m9sgc393bgo5s; path=/; secure; HttpOnly; SameSite=Lax
< Content-Security-Policy: default-src 'none';
< Set-Cookie: __Host-nc_sameSiteCookielax=true; path=/; httponly;secure; expires=Fri, 31-Dec-2100 23:59:59 GMT; SameSite=lax
< Set-Cookie: __Host-nc_sameSiteCookiestrict=true; path=/; httponly;secure; expires=Fri, 31-Dec-2100 23:59:59 GMT; SameSite=strict

* Replaced cookie oclg6vjwjihc="15ij7urrbcqk2m9sgc393bgo5s" for domain nextcloud.benjaminerhart.com, path /, expire 0

< Set-Cookie: oclg6vjwjihc=15ij7urrbcqk2m9sgc393bgo5s; path=/; secure; HttpOnly; SameSite=Lax
< WWW-Authenticate: Basic realm="Nextcloud", charset="UTF-8"
< Content-Length: 414
< Content-Type: application/xml; charset=utf-8


* Received 414 B chunk
* Connection #0 to host nextcloud.benjaminerhart.com left intact
```

Body:

```xml
<?xml version="1.0" encoding="utf-8"?>
<d:error
	xmlns:d="DAV:"
	xmlns:s="http://sabredav.org/ns">
	<s:exception>Sabre\DAV\Exception\NotAuthenticated</s:exception>
	<s:message>No 'Authorization: Basic' header found. Either the client didn't send one, or the server is misconfigured, No 'Authorization: Bearer' header found. Either the client didn't send one, or the server is mis-configured</s:message>
</d:error>
```


So... I really don't see the need for this `https://nextcloud.example.com/yourusername` thingie. Can you point me to the instance which seems to need this?

tladesignz reviewed (changes requested):
Ok, I see why you're struggling with my advice, now.

It seems, Nextcloud does [want to move over people over from `/remote.php/webdav/` to `/remote.php/dav/$username`](https://github.com/nextcloud/server/issues/25867).

However, the first still works with the latest Nextcloud, and it is more compatible. Hence, I suggest to use that, until it is really really removed on the Nextcloud server code. (Which might well be never.)

So it turns out I actually have an account at [nx27277.your-storageshare.de](http://nx27277.your-storageshare.de/).
I just tried it with Save iOS. Works just fine.

Please replicate the behaviour of Save iOS:

- When the user stops editing the WebDAV server URL, automatically add/replace the scheme to match "https" (because we disallow non-encrypted comms) and add /remote.php/webdav/ as path, in case the user didn't add/edit the path part of the URL already.
- Use some URL/URI helper class to ensure a correct URL gets stored.
- When setting up, just connect to the root and see if you get a 200 OK. If true, store space.
- Ensure, all requests get a HTTP-Basic-Authorization header added. Your `OkHttpBaseClient` should already do that, but it might be the case the the used `BasicAuthInterceptor` doesn't get triggered properly, so you might have to do it manually. Implementing Basic Auth yourself is trivial, though.
- Please refactor all variables, method arguments and object properties, so they actually tell by their name what they contain!
Yes, even the SugarORM ones. (Their might be an annotation to cater for the case where SQL column name cannot be inferred automatically from the variable name. If not, you can still make it private and do a proxy property/differently named accessor.) And drop unused ones!

tladesignz commented on app/src/main/java/net/opendasharchive/openarchive/services/webdav/WebDAVLoginActivity.kt:
```diff
@@ -234,89 +223,73 @@ class WebDAVLoginActivity : AppCompatActivity() {
                 try {
                     try {
                         sardine.getQuota(space.host)
                         sardine.list(space.host + "/files/" + space.username + "/")
```
What's this `Sardine` thing doing here? Seems to still use the newer method which requires a username.

pkhivesara replied:
This is used to get the available storage space, but I dont see it being used anywhere. Removed it for now.

tladesignz reviewed (commented):
Where's the code which automatically adds `/remote.php/webdav/`? Shouldn't there be a change commit which replaces `/remote.php/dav/` with `/remote.php/webdav/`?

pkhivesara commented:
> Where's the code which automatically adds `/remote.php/webdav/`? Shouldn't there be a change commit which replaces `/remote.php/dav/` with `/remote.php/webdav/`?

Done! It works fine with ``/dav/`` as well for some reason though
